networkx/algorithms/dag.py

```diff
@@ -1224,36 +1224,190 @@ def dag_to_branching(G):
 @not_implemented_for("undirected")
 @nx._dispatchable
 def compute_v_structures(G):
-    """Iterate through the graph to compute all v-structures.
+    """Yields 3-node tuples that represent the v-structures in `G`.
 
-    V-structures are triples in the directed graph where
-    two parent nodes point to the same child and the two parent nodes
-    are not adjacent.
+    Colliders are triples in the directed acyclic graph (DAG) where two parent nodes
+    point to the same child node. V-structures are colliders where the two parent
+    nodes are not adjacent. In a causal graph setting, the parents do not directly
+    depend on each other, but conditioning on the child node provides an association.
+
+    .. deprecated:: 3.4
+       `compute_v_structures` actually yields colliders. It will be removed in
+       version 3.6. Use `nx.dag.v_structures` or `nx.dag.colliders` instead.
 
     Parameters
     ----------
     G : graph
-        A networkx DiGraph.
+        A networkx `DiGraph`.
 
-    Returns
-    -------
-    vstructs : iterator of tuples
-        The v structures within the graph. Each v structure is a 3-tuple with the
-        parent, collider, and other parent.
+    Yields
+    ------
+    A 3-tuple representation of a v-structure
+        Each v-structure is a 3-tuple with the parent, collider, and other parent.
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If `G` is an undirected graph.
+
+    Examples
+    --------
+    >>> G = nx.DiGraph([(1, 2), (0, 4), (3, 1), (2, 4), (0, 5), (4, 5), (1, 5)])
+    >>> nx.is_directed_acyclic_graph(G)
+    True
+    >>> list(nx.compute_v_structures(G))
+    [(0, 4, 2), (0, 5, 4), (0, 5, 1), (4, 5, 1)]
+
+    See Also
+    --------
+    v_structures
+    colliders
+
+    Notes
+    -----
+    This function was written to be used on DAGs. But it works on cyclic graphs
+    too and since colliders are referred to in the cyclic causal graph literature
+    [2]_ we allow cyclic graphs in this function. It is suggested that you test if
+    your input graph is acyclic as in the example if you want that property.
+
+    References
+    ----------
+    .. [1]  `Pearl's PRIMER,<http://bayes.cs.ucla.edu/PRIMER/primer-ch2.pdf>`_
+            Ch-2 page 50: v-structures def.
+    .. [2] A Hyttinen, P.O. Hoyer, F. Eberhardt, M J ̈arvisalo, (2013)
+           "Discovering cyclic causal models with latent variables:
+           a general SAT-based procedure", UAI'13: Proceedings of the Twenty-Ninth
+           Conference on Uncertainty in Artificial Intelligence, pg 301–310,
+           `doi:10.5555/3023638.3023669 <https://dl.acm.org/doi/10.5555/3023638.3023669>`_
+    """
+    import warnings
+
+    warnings.warn(
+        (
+            "\n\n`compute_v_structures` actually yields colliders. It will be\n"
+            "removed in version 3.6. Use `nx.dag.v_structures` or `nx.dag.colliders`\n"
+            "instead.\n"
+        ),
+        category=DeprecationWarning,
+        stacklevel=5,
```

Comment:
I'm not sure about this.
(see [comment](https://github.com/networkx/networkx/pull/7398#issuecomment-2161060029))

Reply:
I think it depends on how many function calls occur on the way to calling k_core.

When I run nx3.3 in a mamba environment I don't get any message when calling 
`nx.k_core(MG, core_number=cn)`.
But if I define a function: `def f():return nx.k_core(MG, core_number=cn)` and run `f()`, I do see it. If I bury it deeper (`def g():return f()`, etc for h(), hh(), ...) I still see the warning, but the line shown as causing the warning changes. We have decorators so sometimes it is hard to count how many levels to go up. But I don't understand why running colab would change the number of levels.

Reply:
It seems like `stacklevel=2` works for the DeprecationWarning. That shows the line of the function where `k_core` is called as the location of the warning. A stacklevel of 1 shows the line where the warning is created. I believe a stack level of 3 show the line of a function calling a function that calls `k_core`.
Anyway, that should be good enough and we can make an issue to figure out more about this.   :)

Reply:
I think maybe we can use the `skip_file_prefixes` kwarg here. But it was added in v3.12 so if we add it now then the tests running with Python 3.11 and 3.10 will fail. But, we can consider using this in the future.

from the [docs](https://docs.python.org/3/library/warnings.html):

```The skip_file_prefixes keyword argument can be used to indicate which stack
frames are ignored when counting stack levels. This can be useful when you want
the warning to always appear at call sites outside of a package when a constant
stacklevel does not fit all call paths or is otherwise challenging to maintain.
.....
When prefixes are supplied, stacklevel is implicitly overridden to be max(2, stacklevel)
```

Also, I created a [question](https://stackoverflow.com/questions/78620448/inconsistent-behavior-of-warnings-warn) on Stackoverflow about this.

Reply:
One suggestion from the [question](https://stackoverflow.com/questions/78620448/inconsistent-behavior-of-warnings-warn) was to use the `warnings.simplefilter("always")` filter which will cause all warnings to always be triggered. But, I don't think we should add it here as it might override other filters that a user might have used. 

Also, adding https://github.com/ipython/ipython/issues/11207 for reference.

Reply:
I tried re-running `kcores` in colab and I did see the warning the first time I ran the code(with `stacklevel=5`) but not the next time and next to next time. I think I missed it previously. I think we can merge this PR with `stacklevel=5` @dschult. Thanks!

Reply:
Why stacklevel=5 instead of stacklevel=2 as the SOF question/answer suggested?

Reply:
Also, I think the suggestion of using `warnings.simplefilter("always")` is a suggestion for debugging the correct stacklevel more than it is a suggestion for published code. That is, use that to figure out how it works, but then let the user set their own filter.

Reply:
Looking at this again:  It is true that `stacklevel=5` makes this work while stacklevel=2 or 3 or 4 does not. To figure that out I tried adding code like this in place of the deprecationwarning.
```python
for s in range(8):
    warnings.warn(f"stacklevel={s}", category=DeprecationWarning, stacklevel=s)
```
It shows which stacklevel the warning is displayed. Then when i turn on `warnings.simplefilter("always")` it shows the line of code from each stacklevel. It is pretty clear that the dispatchable decorator adds 3 levels. And IPython adds more layers on top of that when calling each cell.

In any case, `stacklevel=5` is the value we want.

Reply:
yes, I was also only getting the deprecation warning displayed with `stacklevel=5` but I didn't know why. I knew it probably had something to do with decorators (`@nx._dispatchable`, `not_implemented_for`) but I wasn't clear about the details of how and where the `stacklevel` was actually getting increased.